### PR TITLE
Add NoisyNet on DQN

### PR DESCRIPTION
**Pull request by MrSyee**
Result

Lunarlander: 
https://app.wandb.ai/khsyee/reinforcement_learning_examples/runs/q8spcy4v?workspace=user-khsyee

Pong:
https://app.wandb.ai/khsyee/reinforcement_learning_examples/runs/9x6de9me?workspace=user-khsyee

### CODE REVIEW

File: algorithms/common/abstract/agent.py [hunk 1]
```diff
@@ -94,6 +94,14 @@ def write_log(self, *args):
     def train(self):
         pass
 
+    @abstractmethod
+    def train_(self):
```
Comment (Curt-Park) [inline algorithms/common/abstract/agent.py]: Need to describe what it is

File: algorithms/common/abstract/agent.py [hunk 2]
```diff
@@ -94,6 +94,14 @@ def write_log(self, *args):
...
+        pass
+
+    @abstractmethod
+    def eval(self):
```
Comment (Curt-Park) [inline algorithms/common/abstract/agent.py]: Need to describe what it is

File: algorithms/common/networks/mlp.py [hunk 3]
```diff
@@ -77,23 +79,25 @@ def __init__(
...
             self.__setattr__("hidden_fc{}".format(i), fc)
             self.hidden_layers.append(fc)
 
         # set output layers
         if self.use_output_layer:
-            self.output_layer = nn.Linear(in_size, output_size)
-            self.output_layer.weight.data.uniform_(-init_w, init_w)
-            self.output_layer.bias.data.uniform_(-init_w, init_w)
+            self.output_layer = self.linear_layer(in_size, output_size)
+            if isinstance(self.linear_layer, nn.Linear):
```
Comment (Curt-Park) [inline algorithms/common/networks/mlp.py]: How about using a function?

File: algorithms/dqn/networks.py [hunk 4]
```diff
@@ -26,6 +27,58 @@ def get_dist_q(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
         dist, q = self.fc_layers.get_dist_q(x)
         return dist, q
 
+    def reset_noise(self):
+        self.fc_layers.reset_noise()
+
+
+class NoisyLinear(nn.Module):
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: Should add descriptions

File: algorithms/dqn/networks.py [hunk 5]
```diff
@@ -26,6 +27,58 @@ def get_dist_q(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
...
+
+    def reset_parameters(self):
+        mu_range = 1 / math.sqrt(self.in_features)
+        self.weight_mu.data.uniform_(-mu_range, mu_range)
+        self.weight_sigma.data.fill_(self.std_init / math.sqrt(self.in_features))
+        self.bias_mu.data.uniform_(-mu_range, mu_range)
+        self.bias_sigma.data.fill_(self.std_init / math.sqrt(self.out_features))
+
+    @staticmethod
+    def _scale_noise(size):
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: why protected?

**Comment by medipixel** [comment]
This pull request **introduces 3 alerts** when merging f5dfad28df3b7c338bb6e8498097f97d8e33d1ca into 2db691321b462209b6179a09312ce5298edfdc7d - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-226d3a54b7596c9cfd7c15b6f0ac0b504b7bb532)

**new alerts:**

* 3 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: examples/lunarlander_v2/dqfd.py [hunk 6]
```diff
@@ -64,6 +67,11 @@ def run(env: gym.Env, args: argparse.Namespace, state_dim: int, action_dim: int)
     def get_fc_model():
         hidden_sizes = [128, 64]
 
+        linear_layer = nn.Linear
```
Comment (Curt-Park) [inline examples/lunarlander_v2/dqfd.py]: why allocate?

File: algorithms/dqn/networks.py (hunk 4 above)
Comment (Curt-Park) [reply algorithms/dqn/networks.py]: class NoisyWrapper(nn.Module):
    def __init__(self, std_init=0.5):
        self.std_init = std_init

    def _call_(self, in_size: int, out_size: int):
        self.noisy_layer = NoisyNet(in_size, out_size, self.std_init)

    def forward(self, x: torch.tensor) -> torch.Tensor:
        reutn self.noisy_layer(x)

**Comment by Curt-Park** [comment]
FYI
https://app.wandb.ai/curt-park/dqn/reports?view=curt-park%2FC51%20%28DQN%2FDQfD%29

**Comment by medipixel** [comment]
This pull request **introduces 3 alerts** when merging f13389ccc5717a49fadaf7417af027e4e9a43936 into 2db691321b462209b6179a09312ce5298edfdc7d - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-516a851618cbf36c1be919ec95e54d19cb1ed54f)

**new alerts:**

* 3 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by MrSyee** [comment]
I make a new pull request after rebase.

**Pull request closed without merging**